// win32-api/dist/lib/struct.js
'use strict';
// windows data types struct for ref-struct module https://github.com/TooTallNate/ref-struct
Object.defineProperty(exports, '__esModule', { value: true });
const W = require('./windef');
/**
 * Struct usage:
 *
 * import * as Struct from 'ref-struct';
 * import {DS} from 'win32-api';
 *
 * const point = new Struct(DS.POINT)();
 * point.x = 100;
 * point.y = 200;
 *
 */
exports.INITCOMMONCONTROLSEX = {
    dwSize: W.DWORD,
    dwICC: W.DWORD,
};
exports.MSG = {
    hwnd: W.HWND,
    message: W.UINT,
    wParam: W.WPARAM,
    lParam: W.LPARAM,
    time: W.DWORD,
    pt: W.POINT,
};
// https://msdn.microsoft.com/en-us/library/windows/desktop/dd162805(v=vs.85).aspx
exports.POINT = {
    x: W.LONG,
    y: W.LONG,
};
exports.PROCESS_BASIC_INFORMATION = {
    Reserved1: W.PVOID,
    PebBaseAddress: W.PVOID,
    Reserved2: W.PVOID,
    UniqueProcessId: W.ULONG_PTR,
    InheritedFromUniqueProcessId: W.PVOID,
};
exports.UNICODE_STRING = {
    Length: W.USHORT,
    MaximumLength: W.USHORT,
    Buffer: W.PWSTR,
};
exports.WINDOWINFO = {
    cbSize: W.DWORD,
    rcWindow: W.RECT,
    rcClient: W.RECT,
    dwStyle: W.DWORD,
    dwExStyle: W.DWORD,
    dwWindowStatus: W.DWORD,
    cxWindowBorders: W.UINT,
    cyWindowBorders: W.UINT,
    atomWindowType: W.ATOM,
    wCreatorVersion: W.WORD,
};
// https://msdn.microsoft.com/zh-cn/library/windows/desktop/ms633577(v=vs.85).aspx
exports.WNDCLASSEX = {
    cbSize: W.UINT,
    style: W.UINT,
    //'lpfnWndProc': ffi.Function('int32', ['pointer', 'uint32', 'int32', 'uint32']) ,
    lpfnWndProc: W.WNDPROC,
    cbClsExtra: W.INT,
    cbWndExtra: W.INT,
    hInstance: W.HINSTANCE,
    hIcon: W.HICON,
    hCursor: W.HCURSOR,
    hbrBackground: W.HBRUSH,
    lpszMenuName: W.LPCTSTR,
    lpszClassName: W.LPCTSTR,
    hIconSm: W.HICON,
};
exports.RECT = {
    left: W.LONG,
    top: W.LONG,
    right: W.LONG,
    bottom: W.LONG,
};
exports._RECT = exports.RECT;
