// win32-api/dist/lib/user32/api.js
'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
const W = require('../windef');
exports.apiDef = {
    CreateWindowExW: [W.HWND, [
            W.DWORD, W.LPCTSTR, W.LPCTSTR, W.DWORD,
            W.INT, W.INT, W.INT, W.INT,
            W.HWND, W.HMENU, W.HINSTANCE, W.LPVOID,
        ]],
    DefWindowProcW: [W.LRESULT, [W.HWND, W.UINT, W.WPARAM, W.LPARAM]],
    DispatchMessageW: [W.LRESULT, [W.LPMSG]],
    EnumThreadWindows: [W.BOOL, [W.DWORD, W.WNDENUMPROC, W.LPARAM]],
    EnumWindows: [W.BOOL, [W.WNDENUMPROC, W.LPARAM]],
    FindWindowExW: [W.HWND, [W.HWND, W.HWND, W.LPCTSTR, W.LPCTSTR]],
    GetAncestor: [W.HWND, [W.HWND, W.UINT]],
    GetClassInfoExW: [W.BOOL, [W.HINSTANCE, W.LPCTSTR, W.LPWNDCLASSEX]],
    GetMessageW: [W.BOOL, [W.LPMSG, W.HWND, W.UINT, W.UINT]],
    GetParent: [W.HWND, [W.HWND]],
    GetWindow: [W.HWND, [W.HWND, W.UINT]],
    GetWindowInfo: [W.BOOL, [W.HWND, W.PWINDOWINFO]],
    GetWindowLongW: [W.LONG, [W.HWND, W.INT]],
    GetWindowTextW: [W.INT, [W.HWND, W.LPTSTR, W.INT]],
    GetWindowThreadProcessId: [W.DWORD, [W.HWND, W.LPDWORD]],
    IsWindowVisible: [W.BOOL, [W.HWND]],
    RegisterClassExW: [W.ATOM, [W.WNDCLASSEX]],
    SetWindowTextW: [W.BOOL, [W.HWND, W.LPCTSTR]],
    SetWinEventHook: [W.HWINEVENTHOOK, [W.UINT, W.UINT, W.HMODULE, W.WINEVENTPROC, W.DWORD, W.DWORD, W.UINT]],
    ShowWindow: [W.BOOL, [W.HWND, W.INT]],
    TranslateMessage: [W.BOOL, [W.LPMSG]],
    TranslateMessageEx: [W.BOOL, [W.LPMSG]],
    UnhookWinEvent: [W.BOOL, [W.HWINEVENTHOOK]],
    UpdateWindow: [W.BOOL, [W.HWND]],
};
if (process.arch === 'x64') {
    exports.apiDef.GetWindowLongPtrW = [W.LONG_PTR, [W.HWND, W.INT]];
}
