// win32-api/dist/lib/user32/constants.js
'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
/* --------- Window Styles ---------------- */
// https://msdn.microsoft.com/en-us/library/windows/desktop/ms632600(v=vs.85).aspx
exports.WS_BORDER = 0x00800000;
exports.WS_CAPTION = 0x00C00000;
exports.WS_CHILD = 0x40000000;
exports.WS_CLIPCHILDREN = 0x02000000;
exports.WS_CLIPSIBLINGS = 0x04000000;
exports.WS_DISABLED = 0x08000000;
exports.WS_DLGFRAME = 0x00400000;
exports.WS_GROUP = 0x00020000;
exports.WS_HSCROLL = 0x00100000;
exports.WS_ICONIC = 0x20000000;
exports.WS_MAXIMIZE = 0x01000000;
exports.WS_MAXIMIZEBOX = 0x00010000;
exports.WS_MINIMIZE = 0x20000000;
exports.WS_MINIMIZEBOX = 0x00020000;
exports.WS_OVERLAPPED = 0x00000000;
exports.WS_POPUP = 0x80000000; // The windows is a pop-up window
exports.WS_SIZEBOX = 0x00040000;
exports.WS_SYSMENU = 0x00080000; // The window has a window menu on its title bar.
exports.WS_TABSTOP = 0x00010000;
exports.WS_THICKFRAME = 0x00040000;
exports.WS_TILED = 0x00000000;
exports.WS_VISIBLE = 0x10000000;
exports.WS_VSCROLL = 0x00200000;
exports.WS_OVERLAPPEDWINDOW = (exports.WS_OVERLAPPED | exports.WS_CAPTION | exports.WS_SYSMENU | exports.WS_THICKFRAME | exports.WS_MINIMIZEBOX | exports.WS_MAXIMIZEBOX);
exports.WS_POPUPWINDOW = (exports.WS_POPUP | exports.WS_BORDER | exports.WS_SYSMENU);
exports.WS_TILEDWINDOW = (exports.WS_OVERLAPPED | exports.WS_CAPTION | exports.WS_SYSMENU | exports.WS_THICKFRAME | exports.WS_MINIMIZEBOX | exports.WS_MAXIMIZEBOX);
/* --------- Extended Window Styles ---------------- */
// https://msdn.microsoft.com/en-us/library/windows/desktop/ff700543(v=vs.85).aspx
exports.WS_EX_ACCEPTFILES = 0x00000010;
exports.WS_EX_APPWINDOW = 0x00040000;
exports.WS_EX_CLIENTEDGE = 0x00000200;
exports.WS_EX_COMPOSITED = 0x02000000;
exports.WS_EX_CONTEXTHELP = 0x00000400;
exports.WS_EX_CONTROLPARENT = 0x00010000;
exports.WS_EX_DLGMODALFRAME = 0x00000001;
exports.WS_EX_LAYERED = 0x00080000;
exports.WS_EX_LAYOUTRTL = 0x00400000;
exports.WS_EX_LEFT = 0x00000000;
exports.WS_EX_LEFTSCROLLBAR = 0x00004000;
exports.WS_EX_LTRREADING = 0x00000000;
exports.WS_EX_MDICHILD = 0x00000040;
exports.WS_EX_NOACTIVATE = 0x08000000;
exports.WS_EX_NOINHERITLAYOUT = 0x00100000;
exports.WS_EX_NOPARENTNOTIFY = 0x00000004;
exports.WS_EX_NOREDIRECTIONBITMAP = 0x00200000;
exports.WS_EX_RIGHT = 0x00001000;
exports.WS_EX_RIGHTSCROLLBAR = 0x00000000;
exports.WS_EX_RTLREADING = 0x00002000;
exports.WS_EX_STATICEDGE = 0x00020000;
exports.WS_EX_TOOLWINDOW = 0x00000080;
exports.WS_EX_TOPMOST = 0x00000008;
exports.WS_EX_TRANSPARENT = 0x00000020;
exports.WS_EX_WINDOWEDGE = 0x00000100;
exports.WS_EX_OVERLAPPEDWINDOW = (exports.WS_EX_WINDOWEDGE | exports.WS_EX_CLIENTEDGE);
exports.WS_EX_PALETTEWINDOW = (exports.WS_EX_WINDOWEDGE | exports.WS_EX_TOOLWINDOW | exports.WS_EX_TOPMOST);
